core/src/apps/cardano/sign_tx.py

```diff
@@ -171,68 +147,39 @@ def _process_outputs(self):
     def _build_witnesses(self, tx_aux_hash: str):
         witnesses = []
         for input in self.inputs:
-            _, node = derive_address_and_node(self.keychain, input.address_n)
-            message = (
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
-            signature = ed25519.sign_ext(
-                node.private_key(), node.private_key_ext(), message
-            )
-            extended_public_key = (
-                remove_ed25519_prefix(node.public_key()) + node.chain_code()
-            )
-            witnesses.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([extended_public_key, signature])),
-                ]
-            )
-
-        return witnesses
-
-    @staticmethod
-    def compute_fee(input_coins_sum: int, outgoing_coins: list, change_coins: list):
-        outgoing_coins_sum = sum(outgoing_coins)
-        change_coins_sum = sum(change_coins)
-
-        return input_coins_sum - outgoing_coins_sum - change_coins_sum
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
+
+            signature = ed25519.sign_ext(node.private_key(), node.private_key_ext(), message)
+
+            public_key = remove_ed25519_prefix(node.public_key())
+
+            witnesses.append([public_key, signature])
+
+        return {0: witnesses}
```

Comment:
This looks odd (what does that 0 correspond to?), what about returning just a list and turning it in to a map where the map is needed?

Type annotation for return value is missing.

And this should perhaps be named input_witnesses and later merged with reward withdrawal witnesses?

And some TODO comment on not working for either Byron or Shelley, since the witnesses are created in a different way?

Reply:
I have made it so that it returns an array and the map is built later. 
I added type annotations to the whole file, although it's not perfect. 
I renamed the function.
I added the todo comment.